sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<InternalKernel>( ... )
+}); // 1. Device Image is compiled and linked into a program and saved in cache
+    // 2. Prepared program is used to enqueue kernel
+
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc1(); }); // Prepared program is used to enqueue kernel
+});
+
+// Library
+SYCL_EXTERNAL void LibFunc1() {
+// ...
+}
+
+```
+In current cache structure the programs map's key consists of four components:
+kernel set id, specialization constants values, the device this program is built
+for, build options id. In this example Id of kernel set where application's
+kernels can be used to access program cache. However when shared library
+defines kernels and these kernels are run by the application unchanged cache
+structure may lead to double compilation of the same code. Example:
+```
+// Application
+SYCL_EXTERNAL void LibFunc();
+
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc(); });  // Device image for library is compiled
+                                          // and linked together with device
+                                          // image for application, i.e.
+                                          // LibFunc1 and ExternalKernel exist
+                                          // in prepared state
+});
+// ...
+EnqueueLibraryKernel(Q); // If cache mechanism is not changed, this line will
+                         // lead to second compilation of ExternalKernel and
+                         // LibFunc1
+
+// Library
+SYCL_EXTERNAL void LibFunc1() {
+// ...
+}
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+```
+Such case can be optimized by bringing nesting into cache keys structure.
+Kernel set id can be found for each kernel using its name and OS module it is
+coming from. In presence of dynamic linking resulting program can be combined
+out of device images which come from different OS modules. So, it should be
+possible to find needed program by kernel name and any OS module that was
+involved in this program. The new mapping structure is:
+```
+{kernel name} =>
+  {OSModuleHandle, spec const, opts, dev} => program
+```
+I.e. each kernel name is mapped to a set of tuples that consists of OS module,
+spec constant values, JIT compiler options, device. Then concrete tuple is
+mapped to a program object.
+Example:
+```
+// Application
+// OSModule = 1
+
+SYCL_EXTERNAL void LibFunc();
+queue.submit(parallel_for<InternalKernel>( ... ));
+
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([] { LibFunc(); });
+});
+
+EnqueueLibraryKernel(q);
+
+// Library
+// OSModule = 2
+
+SYCL_EXTERNAL lib1_func();
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+
+Program cache will have the following structure:
+ "InternalKernel" =>
+   {1, ...} => program 1
+ "ExternalKernel" =>
+   {1, ...} => program 1
+   {2, ...} => program 1
```

Comment:
Need to point that some of the cache items are virtually artificial, i.e. they're prepared at some point.
For instance, one might get confused, how `ExternalKernel => {2, ...} => program 1` gets into cache instead of `ExternalKernel => {2, ...} => program 2`.

Reply:
I've tried to explain it a bit, please see lines 335-342.